### PR TITLE
Validate LoRA truncation admission

### LINKED ISSUE #1528: Harden training admission validators and resolved-control receipts
> *This was generated by AI during triage.*

## Parent

- #1519
- Source milestone: #43
- Related guardrail trackers: #1498, #1499

## What to build

Harden training admission validators so unsupported or internally inconsistent LoRA/QLoRA training requests fail with typed operator errors and resolved-control receipts before training starts.

This slice owns hyperparameter bounds, typed 422 details, resolved defaults, dataset-file receipts, grad clipping, eval batch size, and scheduler compatibility receipts. It should coordinate with #1498/#1499 without duplicating queue or broad trainability UI scope.

## Acceptance criteria

- [ ] Invalid hyperparameters return typed 422-style details with field, reason, received value, and allowed bounds.
- [ ] Legal sentinel modes remain accepted and are reflected in resolved-control receipts.
- [ ] Capability-gated defaults are recorded before worker launch.
- [ ] Dataset file resolution, grad clipping, eval batch size, and scheduler kwargs omission are visible in training receipts.
- [ ] CLI/API/Desktop surfaces show the same admission outcome for the focused fixtures.

## Governing docs

- `docs/runbooks/phase-8-lora-adapter-workflow.md`
- New implementation plan should be created before code changes.
- Related: #1498, #1499

## Expected receipt fields

- `validation_errors`
- `resolved_bounds`
- `capability_gate`
- `dataset_files_resolved`
- `grad_clip_policy`
- `eval_batch_size`
- `scheduler_kwargs_omitted`

## Focused tests

- Invalid hyperparameter matrix tests.
- Legal sentinel mode tests.
- CLI/API typed 422 rendering tests.
- LoRA/QLoRA smoke manifest assertion tests.

## Blocked by

None - can start immediately.

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Add early typed LoRA admission failures when response-only max-length truncation leaves zero unmasked completion tokens.
- Preserve chat media refs/token hints through dataset normalization and reject media-token samples that cannot fit `max_seq_length` before MLX-LM training starts.
- Update the governing LoRA runbook and add the implementation plan for the remaining #1528 truncation-admission slice.

Closes #1528.

## Plan or Spec
- `docs/plans/2026-06-07-issue-1528-truncation-admission.md`
- `docs/runbooks/phase-8-lora-adapter-workflow.md`

## Commands Run
```text
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx pytest -q services/mlx-worker-python/tests/test_lora_model_ops_unit.py::test_run_subprocess_extracts_terminal_structured_result_without_splitlines services/mlx-worker-python/tests/test_lora_model_ops_unit.py::test_mlx_lm_runner_train_native_fails_when_response_only_labels_are_truncated services/mlx-worker-python/tests/test_training_dataset_builder.py::test_write_normalized_dataset_snapshot_applies_manifest_overrides services/mlx-worker-python/tests/test_lora_training_admission.py
# 4 passed in 0.26s after merging origin/main

git diff --cached --check
# pass before commit

git commit -m "Validate LoRA truncation admission"
# pre-commit ran make swift-test, make py-test, make integration-test, and PR-scoped performance report
# make swift-test: pass, elapsed 201.2s
# make py-test: 3633 passed, 14 skipped, 2 warnings in 161.71s
# make integration-test: 117 passed, 1 skipped in 414.09s
# performance report: Status ok, Regressions 0, Verification failures 0

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python python - <<'PY'
from pathlib import Path
import subprocess
import sys
sys.path.insert(0, str(Path.cwd()))
sys.path.insert(0, str(Path.cwd() / 'services/mlx-worker-python'))
from scripts.pre_commit_gate import run_performance_report
root = Path.cwd()
changed = subprocess.check_output(['git', 'diff', '--name-only', 'origin/main', 'HEAD', '--'], cwd=root, text=True).splitlines()
changed = [line.strip() for line in changed if line.strip()]
outcome = run_performance_report(root, changed, base_ref='origin/main')
print(f'outcome_status={outcome.status}')
print(f'report_dir={outcome.report_dir}')
raise SystemExit(0 if outcome.status == 'ok' else 1)
PY
# merge-after-origin/main scoped performance: Status ok, Regressions 0, Verification failures 0

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage run -m pytest -q services/mlx-worker-python/tests/test_lora_model_ops_unit.py::test_run_subprocess_extracts_terminal_structured_result_without_splitlines services/mlx-worker-python/tests/test_lora_model_ops_unit.py::test_run_subprocess_rejects_missing_structured_result services/mlx-worker-python/tests/test_lora_model_ops_unit.py::test_extract_structured_result_payload_accepts_carriage_return_line_end services/mlx-worker-python/tests/test_lora_model_ops_unit.py::test_extract_structured_result_payload_skips_embedded_prefix_and_finds_prior_line services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_mlx_lm_runner_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_mlx_lm_result_tail_probe_script_emits_metrics services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_mlx_lm_result_tail_probe_script_main_covers_checked_in_file
# review-fix probe coverage: 8 passed in 2.24s; changed_scope_coverage 100%, 8/8 changed lines covered

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx pytest -q services/mlx-worker-python/tests/test_lora_model_ops_unit.py::test_run_subprocess_extracts_terminal_structured_result_without_splitlines services/mlx-worker-python/tests/test_lora_model_ops_unit.py::test_mlx_lm_runner_train_native_fails_when_response_only_labels_are_truncated services/mlx-worker-python/tests/test_training_dataset_builder.py::test_write_normalized_dataset_snapshot_applies_manifest_overrides services/mlx-worker-python/tests/test_lora_training_admission.py
# review-fix focused suite: 4 passed in 0.32s

git diff --check
# pass before review-fix commit

git commit -m "Harden response-only admission length fallback"
# pre-commit ran make swift-test, make py-test, make integration-test, and PR-scoped performance report
# make swift-test: pass, elapsed 230.4s
# make py-test: 3633 passed, 14 skipped, 2 warnings in 165.65s
# make integration-test: 117 passed, 1 skipped in 429.08s
# performance report: Status ok, Regressions 0, Verification failures 0
```

## Coverage and Metrics
- Initial focused changed-line coverage before commit: TOTAL 98.41%, 186/189 changed executable lines covered.
- Initial pre-commit scoped performance report: `/Users/chenyu/Documents/github/melix/.runtime/worktrees/issue-1528-truncation-admission-20260607/.runtime/pre-commit-performance/20260607-144649-e381a43b/report/report.md`; Status `ok`, regressions `0`, verification failures `0`.
- Merge-after-origin/main scoped performance report: `/Users/chenyu/Documents/github/melix/.runtime/worktrees/issue-1528-truncation-admission-20260607/.runtime/pre-commit-performance/20260607-145110-477906fd/report/report.md`; Status `ok`, regressions `0`, verification failures `0`.
- Review-fix probe changed-line coverage: TOTAL 100%, 8/8 changed executable lines covered.
- Review-fix pre-commit scoped performance report: `/Users/chenyu/Documents/github/melix/.runtime/worktrees/issue-1528-truncation-admission-20260607/.runtime/pre-commit-performance/20260607-151416-477906fd/report/report.md`; Status `ok`, regressions `0`, verification failures `0`.
- Selected probes: training dataset quality/token summary, validation split partial selection, validation sample-limit loading, MLX-LM structured result tail parse. All direct probe coverage passed; review-fix MLX-LM probe coverage passed at 100.0%.

## Known Gaps
- Real VLM processor image/video tokenization is not re-run in admission; media truncation admission uses dataset-provided media token hints preserved in normalized chat samples.
- No protocol or dependency changes in this PR.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts. N/A: no protocol changes.
- [x] Dependency changes include updated lockfiles. N/A: no dependency changes.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason. Observability mode: minimal typed admission details and existing PR-scoped performance probes; no deferred debug artifacts.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/mlx-worker-python/worker/model_ops/mlx_lm_runner.py [hunk 1]
```diff
@@ -744,6 +745,12 @@ def _validate_response_only_trainable_tokens(
         return
     if aggregate.trainable_response_token_count > 0:
         return
+    affected_sample_count = (
+        aggregate.fully_truncated_response_sample_count
+        if aggregate.fully_truncated_response_sample_count > 0
+        else aggregate.sample_count
+    )
+    suggested_minimum = max(int(aggregate.boundary_max) + 1, int(request.config.max_seq_length) + 1)
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/model_ops/mlx_lm_runner.py]: Fixed in 26ef2d5c: response-only admission now normalizes max_seq_length and boundary_max through the same positive-int fallback used by media admission, and the probe-covered unit test covers a non-integer max_seq_length while preserving typed error details.
